lib/lf/fe/loc_comp_ellbvp.h

```diff
@@ -20,214 +20,236 @@
+ * is either a SCALAR compatible type of a matrix type like
+ * `Eigen::Matrix<SCALAR,...>`.
+ * - REACTION_COEFF must behave like `std::function<SCALAR(Eigen::Vector2d)>`.
+ *
+ * @note This class complies with the type requirements for the template
+ argument
+ * ELEM_MAT_COMP of the function lf::assemble::AssembleMatrixLocally().
+ *
+ * The element matrix is that for the (local) bilinear form
+ * @f[
+    (u,v) \mapsto\int\limits_{K}\boldsymbol{\alpha}(\mathbf{x})\mathbf{grad}\,u
+          \cdot\mathbf{grad}\,v + \gamma(\mathbf{x})u\,v\,\mathrm{d}\mathbf{x}
+ \;,
+ * @f]
+ * with _diffusion coefficient_ @f$\mathbf{\alpha}@f$ and reaction coefficient
+ @f$\gamma@f$.
  */
-template <typename DIFF_COEFF, typename REACTION_COEFF>
-class LagrangeFEEllBVPElementMatrix {
+template <typename SCALAR, typename DIFF_COEFF, typename REACTION_COEFF>
+class LagrangeFEEllBVPElementMatrix : public LocCompLagrFEPreprocessor {
  public:
   /**
    * @brief type of returned element matrix
    */
-  using ElemMat = Eigen::MatrixXd;
-  /*
+  using elem_mat_t = Eigen::Matrix<SCALAR, Eigen::Dynamic, Eigen::Dynamic>;
+  /** @brief Return type for @ref Eval() method */
+  using ElemMat = const elem_mat_t;
+  /** @brief Type for diffusion coefficient */
+  using diff_coeff_t =
+      typename std::invoke_result<DIFF_COEFF, Eigen::VectorXd>::type;
+  /** @brief Type for reaction coefficient */
+  using reac_coeff_t =
+      typename std::invoke_result<REACTION_COEFF, Eigen::VectorXd>::type;
+
+  /** @brief standard constructors */
+  /** @{ */
+  LagrangeFEEllBVPElementMatrix(const LagrangeFEEllBVPElementMatrix &) = delete;
+  LagrangeFEEllBVPElementMatrix(LagrangeFEEllBVPElementMatrix &&) noexcept =
+      default;
+  LagrangeFEEllBVPElementMatrix &operator=(
+      const LagrangeFEEllBVPElementMatrix &) = delete;
+  LagrangeFEEllBVPElementMatrix &operator=(LagrangeFEEllBVPElementMatrix &&) =
+      default;
+  /** @} */
+
+  /**
    * @brief Constructor: cell-independent precomputations
    *
-   * @param fe_trie finite element to be used on triangles
-   * @param fe_quad finite element for quadrilaterals
+   * @param fe_trie_p finite element to be used on triangles
+   * @param fe_quad_p finite element for quadrilaterals
    *
-   * The two parametric finite elements for triangles and quadrilaterals have to
-   * be compatible in the sense that they all assign exactly one reference shape
-   * function to each vertex and the same number of interior shape functions
-   * to each edge
+   * @see LocCompLagrFEPreprocessor::LocCompLagrFEPreprocessor()
    */
   LagrangeFEEllBVPElementMatrix(
-      const ScalarReferenceFiniteElement<double> &fe_tria,
-      const ScalarReferenceFiniteElement<double> &fe_quad, DIFF_COEFF alpha,
-      REACTION_COEFF gamma);
+      std::shared_ptr<const ScalarReferenceFiniteElement<double>> fe_tria_p,
+      std::shared_ptr<const ScalarReferenceFiniteElement<double>> fe_quad_p,
+      DIFF_COEFF alpha, REACTION_COEFF gamma);
```

Comment:
What if `alpha` or `gamma` are functions that are piecewise constant per element of the mesh?

Reply:
fixed in `mesh_function`